client/dashboard/leaderboards/index.js

```diff
@@ -0,0 +1,184 @@
+/** @format */
+/**
+ * External dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { Component, Fragment } from '@wordpress/element';
+import { compose } from '@wordpress/compose';
+import { isEqual } from 'lodash';
+import PropTypes from 'prop-types';
+import { SelectControl, ToggleControl } from '@wordpress/components';
+import { withDispatch } from '@wordpress/data';
+
+/**
+ * WooCommerce dependencies
+ */
+import { EllipsisMenu, MenuItem, MenuTitle, SectionHeader } from '@woocommerce/components';
+
+/**
+ * Internal dependencies
+ */
+import withSelect from 'wc-api/with-select';
+import TopSellingProducts from './top-selling-products';
+import './style.scss';
+
+class Leaderboards extends Component {
+	constructor( props ) {
+		super( ...arguments );
+		this.state = {
+			hiddenLeaderboardKeys: props.userPrefLeaderboards || [],
+			rowsPerTable: props.userPrefLeaderboardRows || 5,
+		};
+
+		this.toggle = this.toggle.bind( this );
+	}
+
+	componentDidUpdate( {
+		userPrefLeaderboardRows: prevUserPrefLeaderboardRows,
+		userPrefLeaderboards: prevUserPrefLeaderboards,
+	} ) {
+		const { userPrefLeaderboardRows, userPrefLeaderboards } = this.props;
+		if ( userPrefLeaderboards && ! isEqual( userPrefLeaderboards, prevUserPrefLeaderboards ) ) {
+			/* eslint-disable react/no-did-update-set-state */
+			this.setState( {
+				hiddenLeaderboardKeys: userPrefLeaderboards,
+			} );
+			/* eslint-enable react/no-did-update-set-state */
+		}
+		if (
+			userPrefLeaderboardRows &&
+			parseInt( userPrefLeaderboardRows ) !== parseInt( prevUserPrefLeaderboardRows )
+		) {
+			/* eslint-disable react/no-did-update-set-state */
+			this.setState( {
+				rowsPerTable: parseInt( userPrefLeaderboardRows ),
+			} );
+			/* eslint-enable react/no-did-update-set-state */
+		}
+	}
+
+	toggle( key ) {
+		return () => {
+			this.setState(
+				prevState => {
+					const hidden = prevState.hiddenLeaderboardKeys.includes( key );
+					const hiddenLeaderboardKeys = prevState.hiddenLeaderboardKeys.filter(
+						chartKey => chartKey !== key
+					);
+
+					if ( ! hidden ) {
+						hiddenLeaderboardKeys.push( key );
+					}
+					return { hiddenLeaderboardKeys };
+				},
+				() => {
+					const userDataFields = {
+						[ 'dashboard_leaderboards' ]: this.state.hiddenLeaderboardKeys,
```

Comment:
Optional refactor: What do you think about using an object instead of an array for the leaderboard preferences? I think that would clean up a lot of the toggling logic.

Reply:
I think that's a great idea.  We can use the lodash `xor` function that @Aljullu recommended [here](https://github.com/woocommerce/wc-admin/pull/1225#discussion_r245731281) to simplify the array logic down to one line.

Added in https://github.com/woocommerce/wc-admin/pull/1234/commits/917f75226c6630967cb9ddd404a5063fd78afd6a.